storage_controller/src/service.rs

```diff
@@ -1162,6 +1162,11 @@ impl Service {
             }
         }
 
+        // If we just finished detaching all shards for a tenant, it might be time to drop it from memory.
+        if tenant.policy == PlacementPolicy::Detached {
+            self.maybe_drop_tenant(tenant.tenant_shard_id.tenant_id, &mut locked);
+        }
```

Comment:
This is a bit scary. Since service state sits behind a sync lock we follow this pattern:
1. acquire lock and get some in-mem state
2. do something async
3. acquire lock again and update something

If step (3) doesn't expect the removal, then we run into trouble. I couldn't find any place with problematic `expect` or `unwrap` calls. Generally, this should be pretty safe since we wait on the reconcile spawned by the detach in `tenant_location_config` and that holds the tenant exclusive lock, but might run into issues if detaches end up taking long.

Reply:
This comment isn't really actionable, but I'm curious about your thoughts on this.

Reply:
Request handlers, the `tenant_op_locks` for the tenant should prevent any shenanigans like this.

However, you make an excellent point in this particular context: process_result does _not_ hold that lock.  If a request handler took the lock, read the existence of the tenant, and then raced with the processing of a reconciler completion, this could violate the assumption that while holding the lock, a tenant that is in memory should remain in memory.

I think the neatest solution to this is to try and get an exclusive lock around this, and to make our maybe_load and maybe_drop functions take refs to lock handles to prove the callers aren't using them outside the lock, let's try that...